packages/binarytree/src/binaryTree.ts

```diff
@@ -0,0 +1,750 @@
+import { type BinaryTreeOpts, ROOT_DB_KEY } from './types.js'
+
+import type { BinaryNode } from './node/types.js'
+import type { PutBatch } from '@ethereumjs/util'
+import type { Debugger } from 'debug'
+
+interface Path {
+  node: BinaryNode | null
+  remaining: number[]
+  stack: Array<[BinaryNode, number[]]>
+}
+
+/**
+ * The basic binary tree interface, use with `import { BinaryTree } from '@ethereumjs/binarytree'`.
+ */
+export class BinaryTree {
+  /** The options for instantiating the binary tree */
+  protected _opts: BinaryTreeOpts
+
+  /** The root for an empty tree */
+  EMPTY_TREE_ROOT: Uint8Array
+
+  protected _db!: CheckpointDB
+  protected _hashLen: number
+  protected _lock = new Lock()
+  protected _root: Uint8Array
+
+  protected DEBUG: boolean
+  protected _debug: Debugger = debug('binarytree:#')
+  protected debug: (...args: any) => void
+  /**
+   * Creates a new binary tree.
+   * @param opts Options for instantiating the binary tree
+   *
+   * Note: in most cases, the static {@link createBinaryTree} constructor should be used. It uses the same API but provides sensible defaults
+   */
+  constructor(opts: BinaryTreeOpts) {
+    this._opts = opts
+
+    if (opts.db instanceof CheckpointDB) {
+      throw new Error('Cannot pass in an instance of CheckpointDB')
+    }
+    this._db = new CheckpointDB({ db: opts.db, cacheSize: opts.cacheSize })
+
+    this.EMPTY_TREE_ROOT = new Uint8Array(32)
+    this._hashLen = 32
+    this._root = this.EMPTY_TREE_ROOT
+
+    if (opts?.root) {
+      this.root(opts.root as any)
+    }
+
+    this.DEBUG =
+      typeof window === 'undefined' ? (process?.env?.DEBUG?.includes('ethjs') ?? false) : false
+    this.debug = this.DEBUG
+      ? (message: string, namespaces: string[] = []) => {
+          let log = this._debug
+          for (const name of namespaces) {
+            log = log.extend(name)
+          }
+          log(message)
+        }
+      : (..._: any) => {}
+
+    this.DEBUG &&
+      this.debug(`Trie created:
+    || Root: ${bytesToHex(this._root)}
+    || Persistent: ${this._opts.useRootPersistence}
+    || CacheSize: ${this._opts.cacheSize}
+    || ----------------`)
+  }
+
+  /**
+   * Gets and/or Sets the current root of the `tree`
+   */
+  root(value?: Uint8Array | null): Uint8Array {
+    if (value !== undefined) {
+      if (value === null) {
+        value = this.EMPTY_TREE_ROOT
+      }
+
+      if (value.length !== this._hashLen) {
+        throw new Error(`Invalid root length. Roots are ${this._hashLen} bytes`)
+      }
+
+      this._root = value
+    }
+
+    return this._root
+  }
+
+  /**
+   * Checks if a given root exists.
+   */
+  async checkRoot(root: Uint8Array): Promise<boolean> {
+    try {
+      const value = await this._db.get(root)
+      return value !== null
+    } catch (error: any) {
+      if (error.message === 'Missing node in DB') {
+        return equalsBytes(root, this.EMPTY_TREE_ROOT)
```

Comment:
Is the idea that `checkRoot(this.EMPTY_TREE_ROOT)` should return true?  as written it will not, because db.get won't throw an error looking for it, but also wont find anything there, so this will return `false`

Reply:
I'm actually not sure here.  I think this might have been copied over from another trie class since I don't know if we use this anywhere.  We can definitely fix though.

Reply:
Anyhow let's keep that change, we can revisit this whole handling once we have more clarity on how to handle deleted values etc.